src/gui/qgsmetadatawidget.h

```diff
@@ -0,0 +1,245 @@
+/***************************************************************************
+                          qgsmetadatawidget.h  -  description
+                             -------------------
+    begin                : 17/05/2017
+    copyright            : (C) 2017 by Etienne Trimaille
+    email                : etienne at kartoza.com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSMETADATAWIDGET_H
+#define QGSMETADATAWIDGET_H
+
+#include "QStandardItemModel"
+#include "QStyledItemDelegate"
+
+#include "qgis_gui.h"
+#include "qgsmaplayer.h"
+#include "qgslayermetadata.h"
+#include "ui_qgsmetadatawidget.h"
+
+/**
+ * \ingroup gui
+ * \class QgsMetadataWidget
+ * \brief A wizard to edit metadata on a map layer.
+ *
+ * \since QGIS 3.0
+ */
+
+class GUI_EXPORT QgsMetadataWidget : public QWidget, private Ui::QgsMetadataWidget
+{
+    Q_OBJECT
+
+  public:
+
+    /**
+     * Constructor for the wizard.
+     */
+    QgsMetadataWidget( QWidget *parent, QgsMapLayer *layer = nullptr );
+    ~QgsMetadataWidget();
+
+    /**
```

Comment:
I think all these methods should be made private - we don't want to expose things like this to the API if there's no requirement for them to be used externally, and in this case I worry that locking these into the API might prove restrictive in future.

Reply:
I agree I'm not sure what should be public. I'm not sure functions related to UI are very useful in the public section.
I moved most of them. Should I move more of them into the private section?

Reply:
Two parsers are used in the `QStyledItemDelegate`. Can I let them public? They doesn't break anything I guess, they are just CSV parser.

Reply:
Sure - leave anything public which needs to be, but hide everything which is just implementation details...